client/state/imports/importer-locks/reducer.js

```diff
@@ -0,0 +1,42 @@
+/** @format */
+
+/**
+ * Internal dependencies
+ */
+import { IMPORTS_IMPORT_LOCK, IMPORTS_IMPORT_UNLOCK } from 'state/action-types';
+
+/**
+ * Track whether an importer is currently locked or unlocked
+ *
+ * A locked importer signifies that it should not be able to change its status
+ * until an action explicitly unlocks it.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+export default ( state = {}, action ) => {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_LOCK:
+			if ( ! action.importerId ) {
+				return state;
+			}
+
+			return {
+				...state,
+				[ action.importerId ]: true,
+			};
+		case IMPORTS_IMPORT_UNLOCK:
+			if ( ! action.importerId ) {
+				return state;
+			}
+
+			return {
+				...state,
+				[ action.importerId ]: false,
+			};
+	}
+
+	return state;
+};
```

Comment:
If we consider dispatching actions that relate to user input instead of focusing on internal details about data and side-effects then our reducer becomes simpler and more semantically matched to the import process.

```js
// client/state/imports/reducer.js
import {
	IMPORTER_CANCEL_IMPORT,
	IMPORTER_START_IMPORT,
} from 'state/action-types';

const importIsLocked = ( state = false, { type } ) => {
	switch ( type ) {
		case IMPORTER_CANCEL_IMPORT:
			return true;

		case IMPORTER_START_IMPORT:
			return false;

		default:
			return state;
	}
};

export const reducer = keyedReducer( 'siteId', combineReducers( {
	importIsLocked,
} ) };
```

Reply:
The current system requires an `importerId` to determine whether it is locked, which is why the state is an object. I don't think it can be expressed quite as simply as this.

Reply:
as a side note, I never write correct code. the intention is to convey the concept.

further, if we have import objects by site and by import id then we can still use the `keyedReducer()` to our gain.

```js
keyedReducer( 'siteId', keyedReducer( 'importerId', combineReducers( {

} ) ) );
```

we just want to review if we think that every action dispatched which corresponds to the import system will have a `siteId` and `importerId` property at the top-level of the action

Reply:
Ok - I think I follow ;)

Quick question - do you prefer rebasing existing commits or new commits for code review amendments? I read that its preferred that things are squashed down, but it's hard to know to what degree.

Reply:
I've not seen any policy on this. If you think someone else has your code checked-out or is working in it then it's best to give a notice that you are rebasing so they can be prepared for that and avoid merge messes. Usually we just add new commits, occasionally squash/rebase to keep a branch up-to-date, and almost always do squash-commits through GitHub's interface when we merge.